### PR TITLE
Factorial

### PR DESCRIPTION
**Pull request by shrutivanw**
CS Fundamentals' assignment to compute the factorial of a given non-negative integer.

### CODE REVIEW

File: lib/factorial.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Computes factorial of the input number and returns it
+def factorial(number)
+  puts "NOT IMPLEMENTED"
+  return number
```
Comment (Hamled) [inline lib/factorial.rb]: How would you feel about doing `raise NotImplementedError` here instead?

Best practice I think is generally to avoid `puts` statements and other things that will print to the console for code being tested, because that output kind of gets randomly interspersed with the test runner's output.

File: specs/factorial_spec.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+require 'minitest/autorun'
+require 'minitest/reporters'
+require_relative '../lib/factorial'
+
+describe "factorial" do
+  describe "basic tests" do
+    describe "test 1" do
```
Comment (Hamled) [inline specs/factorial_spec.rb]: Same as the other repos, we should probably remove the `describe` blocks that are immediate parents of individual tests.

File: specs/factorial_spec.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    describe "test 2" do
+      it "factorial(7) = 5040" do
+        factorial(7).must_equal 5040
+      end
+    end
+  end
+
+  # if the parameter is an object, check for nil
+  describe "nil object input test" do
+    it "nil object is not an integer" do
```
Comment (Hamled) [inline specs/factorial_spec.rb]: I think we could move this nil input test case into the edge cases section.

File: specs/factorial_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  describe "edge cases" do
+    describe "edge case 1" do
+      it "factorial(0) = 1" do
+        factorial(0).must_equal 1
+      end
+    end
+
+    describe "edge case 2" do
+      it "factorial(1) = 1" do
+        factorial(0).must_equal 1
```
Comment (Hamled) [inline specs/factorial_spec.rb]: This should be `factorial(1).must_equal 1`.

File: README.md [hunk 5]
```diff
@@ -1 +1,13 @@
-# factorial
\ No newline at end of file
+# Factorial
+Compute the factorial of a given input non-negative integer.
```
Comment (Hamled) [inline README.md:2]: It might be useful for students to also link to the [Wikipedia article for factorial](https://en.wikipedia.org/wiki/Factorial).

**Review by Hamled — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -1 +1,13 @@
...
+## Exercise
+* Design and implement a method that computes factorial of the input non-negative integer and returns it.
+* Factorial of a number, 'n' is represented as *n!* and is defined as follows:
+   - *0!* = *1*
+   - *n!* = *n * (n-1)!* if *n* > 0
+   -      = *n * (n-1) * (n-2) * ... * 1*
+* For example, 3! = 3 * 2 * 1 = 6. 6! = 720.
+* Share and explain the time and space complexities for your solution.
+* If you think of multiple approaches to solve the problem, implement the solution which minimizes space complexity. Explain the other approaches, and your decision in comments above the code.
+**Note**: Do not use Ruby provided functionality.
```
Comment (Hamled) [inline README.md]: We should have an extra blank line above this so that the note goes into its own paragraph.

It might also be good to specify which specific Ruby methods they should avoid. I'm not sure what specifically that would be, since apparently Ruby's standard library doesn't include a factorial function (it does have [gamma](http://ruby-doc.org/core-2.4.0/Math.html#method-c-gamma), though).

File: lib/factorial.rb (hunk 1 above)
Comment (shrutivanw) [reply lib/factorial.rb]: Done!

File: specs/factorial_spec.rb (hunk 2 above)
Comment (shrutivanw) [reply specs/factorial_spec.rb]: Done.

File: specs/factorial_spec.rb (hunk 3 above)
Comment (shrutivanw) [reply specs/factorial_spec.rb]: Done!

File: specs/factorial_spec.rb (hunk 4 above)
Comment (shrutivanw) [reply specs/factorial_spec.rb]: Whoops! Thanks for catching that!

File: README.md (hunk 6 above)
Comment (shrutivanw) [reply README.md]: Good point! It was a copy-paste remnant from splitting up the exercises. I got rid of the note for this exercise.

File: README.md (hunk 5 above)
Comment (shrutivanw) [reply README.md:2]: Good idea! I added a link for further learning.

**Comment by shrutivanw** [comment]
Feedback incorporated. Thank you @Hamled

**Review by Hamled — APPROVED** [review]
